Hotel 8.

dudarenko-dmitry opened the pull request.

dudarenko-dmitry committed:
Task_8:
1. Created branch Hotel_8.
2. Created annotation @ConfigProperty, @AppComponent, @GetInstance, @Entity, @StartPoint, @StartMethod.
3. Created classes for Dependencies' Injection: ClassCollector, DIContainer, DIApplication
4. Refactored program for loading of Configuration using Annotations.
5. Change logic of UserInterface layer.
6. Create "Entry Point" in DIApplication.java and collect here all logic:
        context.createDIContainer();
        context.injectValuesFromDIContainer();
        loadHotelData(context);
        runApplication(context);
        saveHotelData(context);
        quitHotel();
7. Change logic of Serialization/Deserialization.
8. Delete all un-using/commented code.

dudarenko-dmitry committed:
Task_8:
Created branch Hotel_8.
Created annotation @ConfigProperty

dudarenko-dmitry committed:
Task_8:
Created annotation @ConfigProperty.
Refactored program for loading of Configuration using Annotations.

dudarenko-dmitry commented:
Hi, everyone.
Please, check if I understood Task 1 correctly.

Should I use annotation to load all configurations in one class/module while starting program and than transfer them to appropriate places  via getter? For example
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/4b68237b-3bd8-4c96-aad8-7f7c09fe4f0b)

Or should I place annotation only directly under appropriate field/element? Like this:
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/be742b8b-e754-4c59-89c4-084959fb85ab)

Mark005 commented:
you should put them in the services where they are required

Mark005 commented:
you can scan all subfolders to find classes that require fields to inject values, you can read how spring works it will be easier to understand how exactly to implement it

dudarenko-dmitry committed:
Task_8:
Created annotations AppComponent, GetInstance and classes for Dependencies' Injection: ClassCollector, DIContainer, DI.

dudarenko-dmitry committed:
Task_8:
Classes and fields:
1. marked by annotations AppComponent, GetInstance.
2. refactored for using dependency injection.

need additional refactoring!!!

dudarenko-dmitry committed:
Task_8:
Changed logic of using of annotation @ConfigProperty: all properties' values are injected directly to necessery fields.

Mark005 commented on src/pl/senla/hotel/Main.java:
```diff
@@ -1,28 +1,88 @@
 
-        AppConfiguration appConfiguration = AppConfiguration.getAppConfiguration();
+    public static void main(String[] args) throws IllegalAccessException, InvocationTargetException, NoSuchMethodException, ClassNotFoundException, InstantiationException {
+                // Load configuration using Properties-file
```
remove all commented code

dudarenko-dmitry replied:
ok. ready

Mark005 commented on src/pl/senla/hotel/Main.java:
```diff
@@ -1,28 +1,88 @@
+                new FieldAnnotationsScanner(),
+                new TypeAnnotationsScanner(),
+                new SubTypesScanner());
+        Set<Field> annotatedPropertyFields = reflections.getFieldsAnnotatedWith(ConfigProperty.class);
```
it shouldn't be here  because it's part of our dependency injector

dudarenko-dmitry replied:
ok.
I will move it to separate class AnnotationScanner and:
For Task 8.1 (annotation for Properties) than will send received annotated fields to: 
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/456492ea-713c-4844-a5a0-ed85d00d1c11)

For Task 8.2 (annotations for DI) than will create CONTEXT  using annotated classes and their fields.

Mark005 commented on src/pl/senla/hotel/annotations/config/ConfigPropertyAnnotationLoader.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    public void load(Set<Field> fields) throws IllegalAccessException, InvocationTargetException, InstantiationException {
+        for (Field field : fields) {
```
here scan all classes to find annotated with your annotation

dudarenko-dmitry replied:
I described my idea in reply to above comment (Set<Field> fields has already contain only annotated fields) 
Is it rigth way?

Mark005 commented on src/pl/senla/hotel/repository/RepositoryFacilityCollection.java:
```diff
@@ -1,23 +1,26 @@
     }
 
-    public static Repository<HotelFacility> getRepositoryFacility(){
+    public static Repository<HotelFacility> getSingletonInstance(DataStorage<HotelFacility> priceList){
```
get rid of all singletons in your application, you should have a special context where you keep all created instances and should manage them from this context

dudarenko-dmitry replied:
Thank you. Excellent idea.
I have broken my mind trying to find way how get instances from private constructors or method "getSingletonInstance" with huge amount of other parametrs-singlotones.

Mark005 commented on src/pl/senla/hotel/annotations/config/ConfigPropertyAnnotationLoader.java:
```diff
@@ -0,0 +1,78 @@
+        this.configDirectory = configDirectory;
+    }
+
+    public void load(Set<Field> fields) throws IllegalAccessException, InvocationTargetException, InstantiationException {
```
annotation loader better to integrate into your DI loader to inject not only necessary dependencies but also required config props

dudarenko-dmitry replied:
ok

Mark005 commented on src/pl/senla/hotel/annotations/di/DI.java:
```diff
@@ -0,0 +1,41 @@
+    private static DI injector;
+    private DI(){}
+
+    public static DI getSingletonInstance() {
```
you don't need to create it as a singleton, you can create it with the **new**  keyword, it's ok or make it static as spring boot doing 
![image](https://github.com/dudarenko-dmitry/senla_training/assets/39222770/feba2fab-5c63-4fc0-8d42-a37d4dd8a03d)

dudarenko-dmitry replied:
ok

Mark005 commented on src/pl/senla/hotel/annotations/di/DI.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+import java.util.Map;
+
+@AppComponent
```
why is it here? who will manage it?

Mark005 commented on src/pl/senla/hotel/annotations/di/DIContainer.java:
```diff
@@ -0,0 +1,50 @@
+        singletonInstances.put(beanType, bean);
+    }
+
+    public <T> T getBean(Class<T> beanType) {
```
where do you use this method?

dudarenko-dmitry replied:
it was not used yet.
I'm going to use it after creating Context.

Mark005 commented on src/pl/senla/hotel/annotations/di/DIContainer.java:
```diff
@@ -0,0 +1,50 @@
+        for (Class<?> c : appClasses) {
+            AppComponent annotation = c.getAnnotation(AppComponent.class);
+            if (annotation != null) {
+                Method getSingletonInstance = c.getMethod("getSingletonInstance");
```
create a new object with newInstance and always leave an empty constructor, so after creation, you can inject whatever you want

Mark005 replied:
seems like it will not discover nested components that should be injected

Mark005 commented on src/pl/senla/hotel/annotations/di/DIContainer.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.List;
+import java.util.Map;
+
+@AppComponent
```
who will manage it? it's not necessary here

Mark005 commented on src/pl/senla/hotel/comparators/HotelServicesComparatorByPrice.java:
```diff
@@ -1,23 +1,24 @@
 
 import static pl.senla.hotel.constant.HotelServiceConstant.ERROR_IN_SERVICE_TYPE;
 
 public class HotelServicesComparatorByPrice implements Comparator<HotelService> {
```
to be able to inject fields in it, the class should have AppComponent annotation, or I'm wrong?

dudarenko-dmitry replied:
yes. I'll check such classes

Mark005 commented:
ping me on skype when it will be ready for the next review

dudarenko-dmitry committed:
Task_8:
Changed logic of DIContainer's creating.
Need to change logic of UserInterface layer.

dudarenko-dmitry committed:
Task_8:
Change logic of UserInterface layer.
Add annotations: Entity, StartPoint, StartMethod.

dudarenko-dmitry committed:
Task_8:
1. Create "Entry Point" in DIApplication.java and collect here all logic:
        context.createDIContainer();
        context.injectValuesFromDIContainer();
        loadHotelData(context);
        runApplication(context);
        saveHotelData(context);
        quitHotel();
2. Change logic of Serialization/Deserialization.
3. Delete all un-using/commented code.

dudarenko-dmitry committed:
Merge remote-tracking branch 'origin/Hotel_8' into Hotel_8

Mark005 commented on src/pl/senla/hotel/application/config/Configuration.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface Configuration {
+
+    String setPackageToScan();
```
mb **get** instead of **set**?

dudarenko-dmitry replied:
ok. agree

Mark005 commented on src/pl/senla/hotel/application/properties/ConfigPropertyAnnotationLoader.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public static synchronized ConfigPropertyAnnotationLoader getConfigPropertyAnnotationLoader
+            () {
```
formating

dudarenko-dmitry replied:
ready.

Mark005 commented on src/pl/senla/hotel/application/di/DIApplication.java:
```diff
@@ -0,0 +1,104 @@
+        DIContext context = DIContext.getContext();
+        context.createDIContainer();
+        context.injectValuesFromDIContainer();
+        loadHotelData(context);
```
loadHotelData, saveHotelData methods shouldn't be here, because it's infrastructural classes and they shouldn't know anything about your business operations or something like that, you have two options: 1) hardest one you can create a new annotation like springs post construct to call it when bean is configured to load your data from files 2) easier you can move it to repo layer to default constructor to load data right after bean creation

dudarenko-dmitry replied:
ok. I'll create new Controller, Service and repo layers for entity Hotel and operations Load and Save data during Start/Exit application.

dudarenko-dmitry replied:
I've got one more idea:
move calling **"loadingHoteldata()"** to StartMainMenu
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/670bae03-b59a-447f-b56e-638289a77b18)

and invoke **processor.saveHotelData()** in ExecutorMainMenu while the user choose "EXIT":
![image](https://github.com/dudarenko-dmitry/senla_training/assets/102149674/147f1258-d8f9-4cb0-96a8-97b5676d5629)

Mark005 commented on src/pl/senla/hotel/application/di/DIApplication.java:
```diff
@@ -0,0 +1,104 @@
+        loadHotelData(context);
+        runApplication(context);
+        saveHotelData(context);
+        quitHotel();
```
why do you quit here? as i remember it should be an infinity loop and you should manage it from one of the menu options.

dudarenko-dmitry replied:
I changed logic of MainMenu where was all logic for quit from application and save hotel's data before quit.  

Now my application works while method **runApplication(context)**  is running. This method invokes an endless loop of MainMenu. When the user choose "EXIT" the loop will be terminated, method **runApplication(context)** will be executed and methods  **saveHotelData(context),  quitHotel()** will save data and close application.

I'll return old logic if previous logic was correctly.

Mark005 commented on src/pl/senla/hotel/comparators/RoomReservationsComparatorByGuestName.java:
```diff
@@ -1,18 +1,19 @@
 
 public class RoomReservationsComparatorByGuestName implements Comparator<RoomReservation> {
 
+    @GetInstance(beanName = "RepositoryGuestCollection")
```
not the best approach to use the class name as bean name(how you will manage two beans of one class?), better to add a new field in AppComponent  annotation to specify a name if you need and if it's not specified then you can use class name by default, it's ok by now, but keep it in mind

dudarenko-dmitry replied:
ok.
But we don't need manage "two beans of one class" because all our classes's beans are Singleton.

Mark005 replied:
sometimes require to use two beans of the same class, for example, for two objects with different configs, in spring you will give them aliases (bean names) and will decide which of them you want to use, but for now just keep it in mind

Mark005 commented on src/pl/senla/hotel/comparators/RoomReservationsComparatorByGuestName.java:
```diff
@@ -1,18 +1,19 @@
 
-    public RoomReservationsComparatorByGuestName() {
-        this.repositoryGuest = RepositoryGuestCollection.getRepositoryGuest();
+    public RoomReservationsComparatorByGuestName(Repository<Guest> repositoryGuest) {
```
why do you need a repo here?? it's a casual comparator, let him compare objects without calling repo

dudarenko-dmitry replied:
because task 4.3 was  create "List of guests and their rooms (sort alphabetically)".
Entity RoomReservation has only Guest's id.
Guest's name we can receive only from entity Guest by id from GuestRepository.

dudarenko-dmitry committed:
Task_8:
1. Changed logic of application's Start/Exit and Hotel's data load/save
2. Made appropriate changes in DIApplication, StartMenuMain, ExecutorMain

dudarenko-dmitry committed:
Task_8:
a few small adjustments before merging branches.
